source/controllers/src/impedance/VelocityImpedance.cpp

```diff
@@ -0,0 +1,54 @@
+#include "controllers/impedance/VelocityImpedance.hpp"
+#include "state_representation/robot/JointState.hpp"
+#include "state_representation/robot/JointPositions.hpp"
+#include "state_representation/robot/JointVelocities.hpp"
+#include "state_representation/space/cartesian/CartesianState.hpp"
+#include "state_representation/space/cartesian/CartesianPose.hpp"
+#include "state_representation/space/cartesian/CartesianTwist.hpp"
+
+using namespace state_representation;
+
+namespace controllers::impedance {
+template<class S>
+VelocityImpedance<S>::VelocityImpedance(const Eigen::MatrixXd& stiffness,
+                                        const Eigen::MatrixXd& damping) :
+    Impedance<S>(stiffness, damping, Eigen::MatrixXd::Zero(stiffness.rows(), stiffness.cols())) {}
+
+template VelocityImpedance<CartesianState>::VelocityImpedance(const Eigen::MatrixXd&,
+                                                              const Eigen::MatrixXd&);
+template VelocityImpedance<JointState>::VelocityImpedance(const Eigen::MatrixXd&,
+                                                          const Eigen::MatrixXd&);
+
+template<class S>
+S VelocityImpedance<S>::compute_command(const S&, const S&) {
+  throw exceptions::NotImplementedException(
+      "compute_command(desired_state, feedback_state) not implemented for this input class");
+}
+
+template<>
+CartesianState VelocityImpedance<CartesianState>::compute_command(const CartesianState& desired_state,
+                                                                  const CartesianState& feedback_state) {
+  using namespace std::chrono_literals;
+  // compute the displacement by multiplying the desired twist by the unit time period and add it to the current pose
+  CartesianPose desired_pose= static_cast<CartesianPose>(feedback_state) + 1s * static_cast<CartesianTwist>(desired_state);
+  // set this as the new desired_state keeping the rest of the state values
+  CartesianState integrated_desired_state = desired_state;
+  integrated_desired_state.set_pose(desired_pose.data());
+  // compute the impedance control law normally
+  return this->Impedance<CartesianState>::compute_command(integrated_desired_state, feedback_state);
```

Comment:
In this compute_command, there is a call `this->get_inertia()`. What happens there? As I cannot find a default initialization

Reply:
See my comment below. As desired is simply a velocity state, but integrated, the controller can only act in position and velocity. Hence to remove the effect of the acceleration, inertia is set to 0 in the constructor

Reply:
aaah now I found it haha